blog/_posts/2016-08-22-GSoC2016-Graft.md

```diff
@@ -0,0 +1,128 @@
+networks (graphs with attributes defined on vertices and edges).
+Our change in direction was mainly motivated by:
+
+- The challenges associated with distributed graph computations. [This](http://www.frankmcsherry.org/graph/scalability/cost/2015/01/15/COST.html)
+blog post was an eye opener.
+- Only super-massive graphs, of the order of terabytes or petabytes, require distributed execution. Most useful graphs can be analyzed on a single compute node.
+- Multi-threading is not yet fully established in Julia.
+- [LightGraphs.jl](https://github.com/JuliaGraphs/LightGraphs.jl) already provides fast implementations for most graph algorithms.
+
+So the modified proposal could be summarized as, the development of a package that supports:
+
+* Vertex and edge metadata : Key value pairs for vertices and edges.
+* Vertex labelling : Allow vertices to be referenced, externally, through arbitrary Julia types.
+* SQL like queries for edge data and metadata.
+* Compatibility with `LightGraphs`
+
+# Graft
+`ParallelGraphs` turned out to be a misnomer, since I was moving towards a more general purpose data analysis framework. So I chose the name `Graft`, a kind of abbreviation for Graph Toolkit. The following sections detail `Graft's` features:
+
+## Vertex and Edge Metadata
+Graphs are often representations of real world entities, and the relationships between them. Such entities (and their relationships), often have data attached to them.
+While it's quite straightforward to store vertex data (a simple table will suffice), storing edges and their data is very tricky. The data should be structured on the
+source and target vertices, should support random access and should be vectorized for queries.
+
+At first I tried placing the edge data in a SparseMatrixCSC. This turned out to be a bad idea, because sparse matrices are designed for numeric storage.
+A simpler solution is to store edge metadata in a DataFrame, and have a SparseMatrixCSC map edges onto indices for the DataFrame. This strategy needed a lot less
+code, and the benchmarks were more promising. Mutations such as the addition or removal of vertices and edges become more complicated however.
+
+## Vertex Labelling
+Most graph libraries don't support vertex labelling. It can be very confusing to refer to a vertex by its (often long) integer identifier. It's also
+computationally expensive to use non-integer labels in the implementation of the package (any such implementation would involve dictionaries). There is no reason, however,
+for the user to have to use integer labels externally. Graft supports two modes of vertex labelling. By default, a vertex is identified by its internal identifier. A user
+can assign labels of any arbitrary Julia type to identify vertices, overriding the internal identifiers. This strategy, I feel, makes a reasonable compromise between
+user experience and performance.
+
```

Comment:
What do you think about showing a small example that demonstrates the trade-offs?